apps/condo/domains/miniapp/schema/preprocessors/B2BAppAccess.js

```diff
@@ -0,0 +1,521 @@
+const { getType } = require('@keystonejs/utils')
+const { gql } = require('graphql-tag')
+const { isBoolean, isObject, isEmpty, isFunction, get, isArray, set, isString } = require('lodash')
+const pluralize = require('pluralize')
+
+const { execGqlWithoutAccess } = require('@open-condo/codegen/generate.server.utils')
+const { throwAuthenticationError } = require('@open-condo/keystone/apolloErrorFormatter')
+const { evaluateKeystoneAccessResult } = require('@open-condo/keystone/plugins/utils')
+const { GQL_LIST_SCHEMA_TYPE, find, getById } = require('@open-condo/keystone/schema')
+
+
+const ALL_GENERATED_GQL_QUERIES = new Map()
+
+const ALL_PERMISSION_NAMES = new Set()
+
+const PERMISSION_FIELD = {
+    type: 'Checkbox',
+    defaultValue: false,
+    kmigratorOptions: { default: false },
+}
+
+const READ_ONLY_PERMISSION_FIELD = {
+    ...PERMISSION_FIELD,
+    access: {
+        read: true,
+        create: false,
+        update: false,
+    },
+}
+
+function wrapInCustomAccess (defaultAccess, customAccessFn) {
+    // NOTE: you can use the same object in many places! you don't need to wrap it twice
+    if (!customAccessFn.alreadyProcessedByB2BAppAccessPlugin) customAccessFn.alreadyProcessedByB2BAppAccessPlugin = true
+
+    const type = getType(defaultAccess)
+    if (type === 'Boolean') {
+        // No need to wrap! You already have access, or you should not have it anyway!
+        return defaultAccess
+    } else if (type === 'Function') {
+        // NOTE: to prevent multiple wrapping the same function
+        if (defaultAccess.alreadyProcessedByB2BAppAccessPlugin) return defaultAccess
+        else return customAccessFn
+    } else if (type === 'AsyncFunction') {
+        // NOTE: to prevent multiple wrapping the same function
+        if (defaultAccess.alreadyProcessedByB2BAppAccessPlugin) return defaultAccess
+        else return customAccessFn
+    } else if (type === 'Object') {
+        const newAccess = {}
+        if (typeof defaultAccess.read !== 'undefined') newAccess.read = wrapInCustomAccess(defaultAccess.read, customAccessFn)
+        if (typeof defaultAccess.create !== 'undefined') newAccess.create = wrapInCustomAccess(defaultAccess.create, customAccessFn)
+        if (typeof defaultAccess.update !== 'undefined') newAccess.update = wrapInCustomAccess(defaultAccess.update, customAccessFn)
+        if (typeof defaultAccess.delete !== 'undefined') newAccess.delete = wrapInCustomAccess(defaultAccess.delete, customAccessFn)
```

Comment:
Why should we wrap delete and auth operations? Shouldn't they be restricted?

Reply:
They usually have a boolean value, so they have no effect